qualtran/bloqs/factoring/mod_add.py

```diff
@@ -138,3 +141,103 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
 
     def t_complexity(self) -> 'TComplexity':
         return TComplexity(t=2 * self.bitsize)
+
+
+@frozen
+class MontgomeryModAdd(Bloq):
+    r"""An n-bit modular addition gate.
+
+    This gate is designed to operate on integers in the Montgomery form.
+    Implements |x>|y> => |x>|y + x % p> using $4n$ Toffoli
+    gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+        p: The modulus for the addition.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+        y: A bitsize-sized input/output register (register y above).
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585) Fig 6a and 8
+    """
+
+    bitsize: int
+    p: int
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('x', bitsize=self.bitsize), Register('y', bitsize=self.bitsize)])
```

Comment:
Can we use the new dtypes here?

Reply:
Also, should we support different bitsizes for x and y? Is it nontrivial to add that support?

Reply:
I am going to add a new data type MontgomoryUInt which is an unsigned int but in montgomery form (https://en.wikipedia.org/wiki/Montgomery_modular_multiplication). Anything in MontgomeryUInt can be treated as UInt, but not all UInt are in Montgomery form. All mod arithmetic bloqs in the PsiQuantum paper are designed to operate on integers in montgomery form.

I didn't add different bitsizes for x and y because I'm just following the PsiQuantum paper which proposes subroutines with equal bitsizes. The added complexity of adding different bitsize capabilities would involve making sure Add() and SimpleAddConstant() support different bitsizes which really just boils down to adding it for Add().